packages/api/endpoints/distribution.js

```diff
@@ -85,94 +60,129 @@ function getSignedUrl(s3Client, Bucket, Key, username) {
+function isRedirectRequest(request) {
+  return request.resource === '/redirect';
+}
+
+function isAccessTokenExpired(accessTokenRecord) {
+  return accessTokenRecord.expirationTime < Date.now();
+}
+
+async function handleRedirectRequest(params = {}) {
   const {
-    username
-  } = await earthdataLoginClient.getAccessToken(authorizationCode);
+    accessTokenModel,
+    authClient,
+    distributionUrl,
+    request
+  } = params;
+
+  const { code, state } = request.queryStringParameters;
+
+  const getAccessTokenResponse = await authClient.getAccessToken(code);
 
-  return username;
+  await accessTokenModel.create({
+    accessToken: getAccessTokenResponse.accessToken,
+    expirationTime: getAccessTokenResponse.expirationTime,
+    refreshToken: getAccessTokenResponse.refreshToken,
+    username: getAccessTokenResponse.username
+  });
+
+  return new TemporaryRedirectResponse({
+    location: urljoin(distributionUrl, state),
+    cookies: [
+      new Cookie({
+        key: 'accessToken',
+        value: getAccessTokenResponse.accessToken,
+        expires: new Date(getAccessTokenResponse.expirationTime),
+        httpOnly: true,
+        secure: true
+      })
+    ]
+  });
 }
 
-/**
- * An AWS API Gateway function that either requests authentication,
- * or if authentication is found then redirects to an S3 file for download
- *
- * @param {Object} request - an API Gateway request object
- * @param {EarthdataLoginClient} earthdataLoginClient - an instance of an
- *   EarthdataLoginClient that will be used when authorizing this request
- * @param {AWS.S3} s3Client - an AWS S3 client that will be used when processing
- *   this request
- * @returns {Promise<Object>} an API Gateway response object
- */
-async function handleRequest(request, earthdataLoginClient, s3Client) {
-  const granuleLocation = getGranuleLocationFromRequest(request);
-
-  const authorizationCode = get(request, 'queryStringParameters.code');
-
-  if (authorizationCode) {
-    try {
-      const username = await getUsernameFromAuthorizationCode(
-        earthdataLoginClient,
-        authorizationCode
-      );
-
-      const {
-        Bucket,
-        Key
-      } = getBucketAndKeyFromPathParams(granuleLocation);
-
-      log.info({
-        username,
-        accessDate: Date.now(),
-        bucket: Bucket,
-        file: Key,
-        sourceIp: get(request, 'requestContext.identity.sourceIp')
-      });
-
-      const s3RedirectUrl = getSignedUrl(
-        s3Client,
-        Bucket,
-        Key,
-        username
-      );
-
-      return buildRedirectResponse(s3RedirectUrl);
-    }
-    catch (err) {
-      if (err instanceof OAuth2AuthenticationFailure) {
-        return buildClientErrorResponse('Failed to get EarthData token');
-      }
+async function handleFileRequest(params = {}) {
+  const {
+    accessTokenModel,
+    authClient,
+    request,
+    s3Client
+  } = params;
+
+  const redirectToGetAuthorizationCode = new TemporaryRedirectResponse({
```

Comment:
So this redirects someone to actually login/authorize with Earthdata if the access token from their request (in the cookie) cannot be validated?

Reply:
Correct.